Add binding for agent forwarding

Riatre opened the pull request:
channel_open_request_auth_agent_callback is required for implementing ssh agent forward as unlike X11 forward, there is no other way to establish a forwarding channel.

In libssh:

1. Callback is triggered while handling protocol packets in other libssh call.
2. The callback creates a new channel and prepare for bidirectional forwarding between it and ssh agent.
3. The callback then returns a borrow of the newly created channel for libssh to make reply to the remote side.

However, the callback-based flow does not really fit our Rust binding design: during callback we have SessionHolder locked, so it's really hard to do anything without introducing lock re-entrancy issues, plus that it demands us to return a temporary borrow of something owned by Rust side whose lifetime is tricky to model.

Instead, we try to turn the callback-based style back to something resembling `ssh_channel_accept_x11` by buffering pending channels and let users fetch them later in a saner context.

Riatre commented:
I have a draft for using this to implement ssh agent forward in wezterm: https://github.com/wez/wezterm/pull/5345

wez commented on libssh-rs/src/lib.rs:
```diff
@@ -274,6 +282,55 @@ impl Session {
+                    .last_error()
+                    .unwrap_or_else(|| Error::fatal("ssh_channel_new failed")));
+            }
+            match cb(Channel::new(&sess.outer.upgrade().unwrap(), chan)) {
```
In general, I like to avoid adding `unwrap`s when the calling scope can return a Result.
Can you re-phrase this using the `?` operator?

Riatre replied:
Done.

However an upgrade failure really should not happen here, it means a channel outlived the corresponding session, code looks like this now:

```rs
.ok_or_else(|| Error::fatal("impossible: channel callback outlived session"))?
```

Do you still prefer returning an Error, or should I use `.expect()`?

Riatre replied:
Code removed in latest version.

wez commented on libssh-rs/src/lib.rs:
```diff
@@ -274,6 +282,55 @@ impl Session {
+                    .unwrap_or_else(|| Error::fatal("ssh_channel_new failed")));
+            }
+            match cb(Channel::new(&sess.outer.upgrade().unwrap(), chan)) {
+                // SAFETY: We steal a *mut sys::ssh_channel_struct here and let libssh
```
I think the lifetime management around this is a bit sketchy; we cannot guarantee that this is safe as written; there's no way to guarantee that the callback will return the appropriate Error variant on failure.

As an alternative, how about we expose this as a two-callback API?
* First callback is `fn request_agent_callback(channel: &Channel) -> bool` to approve/deny the forward on a non-owning reference to the channel
* Second callback is triggered only when the first returns true and is something like `fn agent_forward_established(channel: Channel)` which passes ownership.

This way the careful internal grubbing around stuff can be contained to this callback and can't leak out to allow the Channel to be used unsafely.

For the destruction code path, I don't like the idea of partially destroying `Channel` and then allowing its Drop impl to run; it allows the object to exist in a partially destroy state that makes it conceptually harder to reason about.

I think an alternative strategy might look like this, which is a little bit more code, but it is reasonably well contained:

```rust
let channel = Channel::new(&sess.outer.upgrade()?, chan);
if request_agent_callback(&channel) {
   agent_forward_established(channel);
   return;
}

unsafe {
 // Disable callbacks during destruction
 // Maybe this could be moved to a pub(crate) method and shared
 // with the Drop impl?
 sys::ssh_remove_channel_callbacks(
     chan.chan_inner,
     chan._callbacks.as_mut());
 // release it
 sys::ssh_channel_free(chan.chan_inner); 
}
// Bypass the Drop impl
std::mem::forget(chan);
```

Riatre replied:
> but I can't come up with a way to fix this without allowing the Channel to be used unsafely

Hm, maybe we can make a `struct UnlockedChannel` without `Send`, pass that to callbacks, and have a `upgrade()` that consumes it and returns back a normal `Channel`. I can do this refactor in another PR first if you prefer so.

wez replied:
FWIW, I still don't think we can guarantee correct behavior; for example, if either callback panics then Channel::Drop will still run and deadlock.

An alternative approach might be to do something like this, using `Arc<Channel>` as a way to manage the lifetime. The callback would take `Arc<Channel>` rather than any other kind of reference to a channel. 

```rust
let mut channel_holder = None;

let result = std::panic::catch_unwind(|| {
   let channel = Arc::new(Channel::new(..)));
   channel_holder.replace(Arc::clone(&channel));
   cb(channel)
});

fn drop_if_only_ref(channel: Arc<Channel>) {
   if let Some(channel) = channel.into_inner() {
      drop_channel_unlocked(channel);
   }
}

match result {
  Err(err) => {
     drop_if_only_ref(channel_holder)
     std::ptr::null_mut()
  }
  Ok(Err(err)) => {
     drop_if_only_ref(channel_holder)
     std::ptr::null_mut()
  }
  Ok(Ok(chan)) => chan,
}
```

The big downside is that this would be the only API surface where Arc is a thing, which feels like an ergonomics wart, and I don't really like it.

Another option is again to split the API surface for this:

* callback to approve/reject the forward as I suggested earlier, accepting `&Channel`
* On approval, the channel is queued to a list of pending channels that is stored in `SessionHolder`.  This can be a `Vec<LockedChannel>` where `LockedChannel` is just the bits we need to remember to free up, which we move from the temporary `Channel` that we pass to the callback.  (Note that we may still need a variant of the above, to safely destruct it across a panic, but we can do so without an Arc):

```rust
struct LockedChannel {
  chan_inner: sys::ssh_channel,
  _callbacks: Box<sys::ssh_channel_callbacks_struct>,
  callback_state: Box<CallbackState>,
}
```

* Add `Session::pop_auth_agent_channel` method to retrieve and move a `LockedChannel` instance and convert it into a `Channel`.  In this context, there is no error path that can trigger and cause a deadlock on drop.

A consequence of this is that the application loop would need to periodically call this new method to detect and process these new channels, which isn't ideal, but is something that can be rolled into wezterm-ssh's SessionEvent reasonably well.

Riatre replied:
I like the `Session::pop_auth_agent_channel` idea! I've just pushed a new version and it actually simplified both libssh-rs and wezterm-ssh side. It largely follows your idea, besides:

> callback to approve/reject the forward as I suggested earlier, accepting `&Channel`

We can't pass the callback a `&Channel`, it still can't do anything on it due to lock issues. Instead, the callback became rather dumb: `fn() -> bool`. API-wise, do we still need a callback here, or should we do `Session::enable_accept_agent_forward` and `Session::disable_accept_agent_forward`?

The only downside of not having a callback I can think of is, now we can't connect to ssh agent before accepting the forwarding request, so in case we can't reach to the agent, we accept the forwarding request and then immediately send EOF instead of refusing it. Should not really matter. WDYT?

wez replied:
> API-wise, do we still need a callback here, or should we do `Session::enable_accept_agent_forward` and `Session::disable_accept_agent_forward`?

Ah. I'd assumed that there was something that the embedding application might want to do to check the channel, but there isn't really anything in the API for it call, so, yeah, I think this whole thing is a lot simpler if we just add a method to set a boolean and we can do away with the actual callback part of this.

So, the question is: do we default to enable or disable?  How does this relate to the corresponding ssh config option?
Does libssh even care about that?
If not, then I think the logical thing to do in the embedding application is to enable/disable based on the config.

Riatre replied:
Done. This is now a trivial 50 lines change. Thanks for your guidance!

> Does libssh even care about that?

No. Yeah, just let embedding app control it.

wez replied:
I like this a lot, thank you for working through this!

wez commented on libssh-rs/src/channel.rs:
```diff
@@ -46,6 +46,9 @@ impl Drop for Channel {
             // Prevent any callbacks firing as part the remainder of this drop operation
             sys::ssh_remove_channel_callbacks(self.chan_inner, self._callbacks.as_mut());
         }
+        if self.chan_inner.is_null() {
```
this should be ordered ahead of the `sys::ssh_remove_channel_callbacks` call above.

However, I think this may not be needed if `forget` is used in the alternative destruction path.

Riatre replied:
Removed in favor of `mem::forget`.

wez commented on libssh-rs/src/error.rs:
```diff
@@ -53,3 +57,9 @@ impl From<std::ffi::NulError> for Error {
+
+impl std::fmt::Debug for RequestAuthAgentError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "RequestAuthAgentError({})", self.0)
```
I think this will go away as part of my other suggestions, but I wanted to note that this is more correctly written as `f.debug_struct("RequestAuthAgentError").field(&self.0).finish()`

Riatre replied:
Done, thanks!

wez commented on libssh-rs/src/lib.rs:
```diff
@@ -72,9 +72,12 @@ fn initialize() -> SshResult<()> {
 }
 
 pub(crate) struct SessionHolder {
+    outer: Weak<Mutex<SessionHolder>>,
```
I'd lean towards naming this something like `self_ref`, and add a comment about how it is needed for callbacks that need to construct a Channel

Riatre replied:
Done.

wez reviewed (changes requested):
Thanks for this! I'd love to be able to use agent forwarding with wezterm!
I think there are some smallish tweaks that are needed to make this safe

Riatre commented:
Thanks for the review! I've addressed most comments, but have further comments on the safety discussion.

In addition, I wonder why do we need `std::panic::catch_unwind` in `bridge_*_callback`-s? I just followed suit here.

wez commented:
Thank you!

tomtastic commented:
I'd also love to see agent forwarding in WezTerm, but small side-note that the Windows OpenSSH agent (and specifically `ssh-add`) [is broken currently for users of SSH certificates](https://github.com/PowerShell/openssh-portable/issues/731).
